st2common/st2common/metrics/metrics.py

```diff
@@ -0,0 +1,162 @@
+
+        if key:
+            assert isinstance(key, str)
+            self._metrics.time(key, time_delta.total_seconds())
+        else:
+            self._metrics.time(self.key, time_delta.total_seconds())
+
+    def get_time_delta(self):
+        """ Get current time delta.
+        """
+        return get_datetime_utc_now() - self._start_time
+
+    def __enter__(self):
+        self._start_time = get_datetime_utc_now()
+        return self
+
+    def __exit__(self, *args):
+        self.send_time()
+
+    def __call__(self, func):
+        @wraps(func)
+        def wrapper(*args, **kw):
+            with self as metrics_timer:
+                return func(*args, metrics_timer=metrics_timer, **kw)
+        return wrapper
+
+
+class Counter(object):
+    """ Timer context manager for easily sending timer statistics.
+    """
+    def __init__(self, key):
+        assert isinstance(key, str)
+        assert key
+        self.key = key
+        self._metrics = METRICS
+
+    def __enter__(self):
+        self._metrics.inc_counter(self.key)
+        return self
+
+    def __exit__(self, *args):
+        self._metrics.dec_counter(self.key)
+
+    def __call__(self, func):
+        @wraps(func)
+        def wrapper(*args, **kw):
+            with self:
+                return func(*args, **kw)
+        return wrapper
+
+
+class CounterWithTimer(object):
+    """ Timer and counter context manager for easily sending timer statistics
+    with builtin timer.
+    """
+    def __init__(self, key):
+        assert isinstance(key, str)
+        assert key
+        self.key = key
+        self._metrics = METRICS
+        self._start_time = None
+
+    def send_time(self, key=None):
+        """ Send current time from start time.
+        """
+        time_delta = get_datetime_utc_now() - self._start_time
+
+        if key:
+            assert isinstance(key, str)
+            self._metrics.time(key, time_delta.total_seconds())
+        else:
+            self._metrics.time("%s%s" % (self.key, METRICS_TIMER_SUFFIX),
+                               time_delta.total_seconds())
+
+    def get_time_delta(self):
+        """ Get current time delta.
+        """
+        return get_datetime_utc_now() - self._start_time
+
+    def __enter__(self):
+        self._metrics.inc_counter("%s%s" % (self.key, METRICS_COUNTER_SUFFIX))
+        self._start_time = get_datetime_utc_now()
+        return self
+
+    def __exit__(self, *args):
+        self.send_time()
+        self._metrics.dec_counter("%s_counter" % self.key)
+
+    def __call__(self, func):
+        @wraps(func)
+        def wrapper(*args, **kw):
+            with self as counter_with_timer:
+                return func(*args, metrics_counter_with_timer=counter_with_timer, **kw)
+        return wrapper
+
+
+try:
+    METRICS = get_plugin_instance(PLUGIN_NAMESPACE, cfg.CONF.metrics.driver)()
+except (NoMatches, MultipleMatches, NoSuchOptError):
+    METRICS = BaseMetricsDriver()
```

Comment:
Seems wrong to put this block of code here. Maybe I don't understand so please clarify. The plugins depends on this module before the plugins are registered, so loading this module will always set METRICS to the BaseMetricsDriver.  Can we move the class definition for BaseMetricsDriver somewhere else and then have the statsd and prometheus plugins import that new file instead of this file?

Reply:
AS we discussed in slack: This isn't used by the driver that imports base. It's loaded at runtime after the drivers have already be registered and the metrics constant will be loaded prior to the driver being registered and put into runtime memory.

Reply:
So if I do the following, METRICS is type of BaseMetricsDriver.  Although the plugins don't use this METRICS constant. The Timer, Counter, TimerWithCounter uses METRICS and so by default it's the BaseMetricsDriver. 

```
>>> from st2common.metrics import metrics
>>> metrics.METRICS
<st2common.metrics.metrics.BaseMetricsDriver object at 0x7fcfcf560bd0>
```

I tried to run st2common's setup.py to see if I can get METRICS to be loaded with the statsd driver. But I'm getting the following errors on setup.

```
Installed /home/vagrant/tmp/st2/virtualenv/lib/python2.7/site-packages/tooz-1.57.4-py2.7.egg
Searching for semver==2.7.7
Reading https://pypi.python.org/simple/semver/
Best match: semver 2.7.7
Downloading https://files.pythonhosted.org/packages/a3/30/dc388e483165014d03e34b9c4cd5e0d5d30794d396225d0cb78b59d86d95/semver-2.7.7.tar.gz#sha256=20ffbb50248a6141bb9eba907db0e47ee4a239ddb55fe0ada8696fc241495a9d
Processing semver-2.7.7.tar.gz
Writing /tmp/easy_install-VsR63w/semver-2.7.7/setup.cfg
Running semver-2.7.7/setup.py -q bdist_egg --dist-dir /tmp/easy_install-VsR63w/semver-2.7.7/egg-dist-tmp-GdRI5C
Traceback (most recent call last):
  File "setup.py", line 68, in <module>
    'statsd = st2common.metrics.drivers.statsd_driver:StatsdDriver',
  File "/usr/lib/python2.7/distutils/core.py", line 151, in setup
    dist.run_commands()
  File "/usr/lib/python2.7/distutils/dist.py", line 953, in run_commands
    self.run_command(cmd)
  File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
    cmd_obj.run()
  File "/home/vagrant/tmp/st2/virtualenv/local/lib/python2.7/site-packages/setuptools/command/develop.py", line 34, in run
    self.install_for_development()
  File "/home/vagrant/tmp/st2/virtualenv/local/lib/python2.7/site-packages/setuptools/command/develop.py", line 133, in install_for_development
    self.process_distribution(None, self.dist, not self.no_deps)
  File "/home/vagrant/tmp/st2/virtualenv/local/lib/python2.7/site-packages/setuptools/command/easy_install.py", line 742, in process_distribution
    [requirement], self.local_index, self.easy_install
  File "/home/vagrant/tmp/st2/virtualenv/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 826, in resolve
    dist = best[req.key] = env.best_match(req, ws, installer)
  File "/home/vagrant/tmp/st2/virtualenv/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 1092, in best_match
    return self.obtain(req, installer)
  File "/home/vagrant/tmp/st2/virtualenv/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 1104, in obtain
    return installer(requirement)
  File "/home/vagrant/tmp/st2/virtualenv/local/lib/python2.7/site-packages/setuptools/command/easy_install.py", line 665, in easy_install
    return self.install_item(spec, dist.location, tmpdir, deps)
  File "/home/vagrant/tmp/st2/virtualenv/local/lib/python2.7/site-packages/setuptools/command/easy_install.py", line 695, in install_item
    dists = self.install_eggs(spec, download, tmpdir)
  File "/home/vagrant/tmp/st2/virtualenv/local/lib/python2.7/site-packages/setuptools/command/easy_install.py", line 876, in install_eggs
    return self.build_and_install(setup_script, setup_base)
  File "/home/vagrant/tmp/st2/virtualenv/local/lib/python2.7/site-packages/setuptools/command/easy_install.py", line 1115, in build_and_install
    self.run_setup(setup_script, setup_base, args)
  File "/home/vagrant/tmp/st2/virtualenv/local/lib/python2.7/site-packages/setuptools/command/easy_install.py", line 1101, in run_setup
    run_setup(setup_script, args)
  File "/home/vagrant/tmp/st2/virtualenv/local/lib/python2.7/site-packages/setuptools/sandbox.py", line 233, in run_setup
    with setup_context(setup_dir):
  File "/usr/lib/python2.7/contextlib.py", line 17, in __enter__
    return self.gen.next()
  File "/home/vagrant/tmp/st2/virtualenv/local/lib/python2.7/site-packages/setuptools/sandbox.py", line 195, in setup_context
    yield
  File "/usr/lib/python2.7/contextlib.py", line 35, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/vagrant/tmp/st2/virtualenv/local/lib/python2.7/site-packages/setuptools/sandbox.py", line 166, in save_modules
    saved_exc.resume()
  File "/home/vagrant/tmp/st2/virtualenv/local/lib/python2.7/site-packages/setuptools/sandbox.py", line 141, in resume
    six.reraise(type, exc, self._tb)
  File "/home/vagrant/tmp/st2/virtualenv/local/lib/python2.7/site-packages/setuptools/sandbox.py", line 154, in save_modules
    yield saved
  File "/home/vagrant/tmp/st2/virtualenv/local/lib/python2.7/site-packages/setuptools/sandbox.py", line 194, in setup_context
    __import__('setuptools')
  File "/home/vagrant/tmp/st2/virtualenv/local/lib/python2.7/site-packages/setuptools/__init__.py", line 11, in <module>
    from setuptools.extern.six.moves import filterfalse, map
  File "/home/vagrant/tmp/st2/virtualenv/local/lib/python2.7/site-packages/setuptools/extern/__init__.py", line 1, in <module>
    from pkg_resources.extern import VendorImporter
  File "/home/vagrant/tmp/st2/virtualenv/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 71, in <module>
    __import__('pkg_resources.extern.packaging.requirements')
  File "/home/vagrant/tmp/st2/virtualenv/local/lib/python2.7/site-packages/pkg_resources/extern/__init__.py", line 42, in load_module
    __import__(extant)
  File "/home/vagrant/tmp/st2/virtualenv/local/lib/python2.7/site-packages/pkg_resources/_vendor/packaging/requirements.py", line 9, in <module>
    from pkg_resources.extern.pyparsing import stringStart, stringEnd, originalTextFor, ParseException
  File "/home/vagrant/tmp/st2/virtualenv/local/lib/python2.7/site-packages/pkg_resources/extern/__init__.py", line 42, in load_module
    __import__(extant)
  File "/home/vagrant/tmp/st2/virtualenv/local/lib/python2.7/site-packages/pkg_resources/_vendor/pyparsing.py", line 696, in <module>
    collections.MutableMapping.register(ParseResults)
  File "/home/vagrant/tmp/st2/virtualenv/lib/python2.7/abc.py", line 109, in register
    if issubclass(subclass, cls):
  File "/home/vagrant/tmp/st2/virtualenv/lib/python2.7/abc.py", line 180, in __subclasscheck__
    if issubclass(subclass, scls):
  File "/home/vagrant/tmp/st2/virtualenv/local/lib/python2.7/site-packages/typing.py", line 1249, in __subclasscheck__
    return super(GenericMeta, self).__subclasscheck__(cls)
TypeError: super() argument 1 must be type, not None
```

Reply:
So I try this differently. Using the get_available_plugins, the statsd driver is there and I get use get_plugin_instance to load the statsd driver. However, when I load the metrics module, the METRICS constant is assigned the BaseMetricsDriver. This looks wrong.  Isn't it suppose to be assigned the StatsdDriver?

```
>>> from st2common.util import loader
>>> loader.get_available_plugins('st2common.metrics.driver')
['statsd']
>>> loader.get_plugin_instance('st2common.metrics.driver', 'statsd')
<class 'st2common.metrics.drivers.statsd_driver.StatsdDriver'>
>>> from st2common.metrics import metrics
>>> metrics.METRICS
<st2common.metrics.metrics.BaseMetricsDriver object at 0x7fc7c2b84450>
```

Reply:
I have similar concerns as @m4dcoder here - we need some kind of init phase or similar, otherwise if we do it directly top-level inside a module, the value will depend on the import order which is something we want to avoid where possible (ideally import should have no side affects).

Reply:
@m4dcoder likely you didn’t import config. Try/except block made it silently return the base.

Reply:
@kami none of this, as far as I can tell at the moment, requires a specific import order. Just that the installation install st2common to register the plugin.